### PR TITLE
Integrate libtty into pc-tty

### PR DESCRIPTION
**Pull request by Maxez**
Changes based on pc-uart's usage of libtty. Fields ttypc_virt_t::m_echo and ttypc_virt_t::m_mln were removed since echo and feed line termios options are handled by libtty.

### CODE REVIEW

File: tty/pc-tty/ttypc_virt.c [hunk 1]
```diff
@@ -128,508 +152,490 @@ static void _ttypc_virt_scroll(ttypc_virt_t *virt)
...
+			break;
+		case 0x0f:  /* SI */
+			virt->GL = &virt->G0;
+			break;
 
-				if (virt->col) {
-					virt->cur_offset--;
-					virt->col--;
-				}
+		case 0x10:  /* DLE */
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_virt.c]: You can compact all cases with _break;_ ending to the one block

File: tty/pc-tty/ttypc_virt.h [hunk 2]
```diff
@@ -16,13 +16,15 @@
...
+#include <sys/types.h>
 
-#include "ttypc.h"
+#include <libtty.h>
 
 
 /* escape detection state machine */
-#define STATE_INIT   0  /* normal	*/
+#define STATE_INIT   0  /* normal */
 #define	STATE_ESC    1  /* got ESC */
```
Comment (agkaminski) [inline tty/pc-tty/ttypc_virt.h]: Wild \t before STATE_ESC

File: tty/pc-tty/ttypc.c [hunk 3]
```diff
@@ -19,7 +19,10 @@
 #include <stdlib.h>
```
[the remark this answers is not in the document]
Comment (Maxez) [reply tty/pc-tty/ttypc.c]: Done.

File: tty/pc-tty/ttypc.c [hunk 4]
```diff
@@ -32,91 +35,112 @@
...
-		return -EINVAL;
+	in_data = ioctl_unpack(msg, &request, &oid.id);
+	out_data = NULL;
+	pid = ioctl_getSenderPid(msg);
 
-	return 0;
+	d = ttypc_virt_get(&oid);
+
+	if (d >= sizeof(ttypc_common.virtuals) / sizeof(ttypc_virt_t))
+		err = -EINVAL;
```
[the remark this answers is not in the document]
Comment (Maxez) [reply tty/pc-tty/ttypc.c]: Fixed.

File: tty/pc-tty/ttypc.c [hunk 5]
```diff
@@ -32,91 +35,112 @@
...
+	memset(&ttypc_common, 0, sizeof(ttypc_t));
+	ttypc_common.color = (inb((void *)0x3cc) & 0x01);
+
+	ttypc_common.out_base = mmap(NULL, 0x1000, PROT_READ | PROT_WRITE, 0, OID_PHYSMEM, ttypc_common.color ? 0xb8000 : 0xb0000);
+	ttypc_common.out_crtc = ttypc_common.color ? (void *)0x3d4 : (void *)0x3b4;
+
+	/* Initialize virutal terminals */
+	for (i = 0; i < sizeof(ttypc_common.virtuals) / sizeof(ttypc_virt_t); i++) {
+		if (_ttypc_virt_init(&ttypc_common.virtuals[i], _PAGE_SIZE, &ttypc_common) < 0) {
+			printf("ttypc: Can't initialize virtual terminal %d!\n", i);
```
[the remark this answers is not in the document]
Comment (Maxez) [reply tty/pc-tty/ttypc.c]: Yes. ttypc port is created after the virtual terminal initialization.

File: tty/pc-tty/ttypc.h [hunk 6]
```diff
@@ -3,9 +3,9 @@
  *
  * ttypc VT220 emulator (based on FreeBSD 4.4 pcvt)
  *
- * Virtual consoel implementation
+ * Virtual console implementation
  *
- * Copyright 217 Pawel Pisarczyk
+ * Copyright 2017 Pawel Pisarczyk
```
[the remark this answers is not in the document]
Comment (Maxez) [reply tty/pc-tty/ttypc.h]: Done.

File: tty/pc-tty/ttypc_kbd.c [hunk 7]
```diff
@@ -327,70 +323,56 @@ static int ttypc_kbd_interrupt(unsigned int n, void *arg)
...
 				ttypc_vga_switch(&ttypc->virtuals[1]);
-			else if (!strcmp((char *)s, "\033[m"))
+			else if (!strcmp(s, "\033[m"))
 				ttypc_vga_switch(&ttypc->virtuals[2]);
-			else if (!strcmp((char *)s, "\033[n"))
+			else if (!strcmp(s, "\033[n"))
 				ttypc_vga_switch(&ttypc->virtuals[3]);
-			else
-				ttypc_virt_sadd(ttypc->cv, s, strlen((char *)s));
+			else {
```
[the remark this answers is not in the document]
Comment (Maxez) [reply tty/pc-tty/ttypc_kbd.c]: Fixed.

File: tty/pc-uart/pc-uart.c [hunk 8]
```diff
@@ -58,7 +58,7 @@ static int uart_interrupt(unsigned int n, void *arg)
 }
```
[the remark this answers is not in the document]
Comment (Maxez) [reply tty/pc-uart/pc-uart.c]: Changed.

File: tty/pc-tty/ttypc.c [hunk 9]
```diff
@@ -126,79 +150,56 @@ void poolthr(void *arg)
...
-
 	portCreate(&port);
-	toid.port = port;
-	if (portRegister(port, "/dev/tty0", &toid) < 0) {
+	if (portRegister(port, "/dev/tty0", &ttypc_common.virtuals[0].oid) < 0) {
 		printf("Can't register port %d\n", port);
 		return -1;
 	}
 
+	void *stack = malloc(2048);
```
[the remark this answers is not in the document]
Comment (Maxez) [reply tty/pc-tty/ttypc.c]: Fixed.

File: tty/pc-tty/ttypc.c [hunk 10]
```diff
@@ -126,79 +150,56 @@ void poolthr(void *arg)
...
+
 	/* Run threads */
-	stack = malloc(2048);
 	beginthread(poolthr, 1, stack, 2048, (void *)port);
```
[the remark this answers is not in the document]
Comment (Maxez) [reply tty/pc-tty/ttypc.c]: Fixed.

File: tty/pc-tty/ttypc_kbd.c [hunk 11]
```diff
@@ -327,70 +323,56 @@ static int ttypc_kbd_interrupt(unsigned int n, void *arg)
...
-	if ((ttypc->rbuff = (uint8_t **)malloc(sizeof(uint8_t *) * 128)) == NULL)
-		return -ENOMEM;
-
-	ttypc->rbuffsz = 128;
-	ttypc->rb = 0;
-	ttypc->rp = 0;
+	void *stack;
 
 	/* Attach interrupt and launch interrupt thread */
 	if ((stack = malloc(2048)) == NULL)
```
[the remark this answers is not in the document]
Comment (Maxez) [reply tty/pc-tty/ttypc_kbd.c]: Fixed.

File: tty/pc-tty/ttypc_virt.c [hunk 12]
```diff
@@ -128,508 +152,490 @@ static void _ttypc_virt_scroll(ttypc_virt_t *virt)
...
 
-/* Function put char ch to the screen according to video state given by svsp */
-static void _ttypc_virt_writechar(ttypc_virt_t *virt, uint16_t attrib, uint16_t ch)
+/* Function put char c to the screen according to video state given by svsp */
+static void _ttypc_virt_writechar(ttypc_virt_t *virt, uint16_t attrib, uint16_t c)
 {
-
-	if ((ch >= 0x20) && (ch <= 0x7f))	{           /* use GL if ch >= 0x20 */
+	if ((c >= 0x20) && (c <= 0x7f)) {               /* use GL if ch >= 0x20 */
 		if (!virt->ss)                              /* single shift G2/G3 -> GL ? */
```
[the remark this answers is not in the document]
Comment (Maxez) [reply tty/pc-tty/ttypc_virt.c]: Fixed.

File: tty/pc-tty/ttypc_virt.c (hunk 1 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_virt.c]: I'd like to leave it 'as is' and maintain case enumeration order.

File: tty/pc-tty/ttypc_virt.h (hunk 2 above)
Comment (Maxez) [reply tty/pc-tty/ttypc_virt.h]: Fixed.

**Pull request merged**